packages/experiment-tag/src/experiment.ts

```diff
@@ -4,461 +4,660 @@ import {
-let urlExposureCache: { [url: string]: { [key: string]: string | undefined } };
-
-export const initializeExperiment = async (
-  apiKey: string,
-  initialFlags: string,
-  config: ExperimentConfig = {},
-) => {
-  const globalScope = getGlobalScope();
-  if (globalScope?.webExperiment) {
-    return;
-  }
-  WindowMessenger.setup();
-  if (!isLocalStorageAvailable() || !globalScope) {
-    return;
-  }
-  previousUrl = undefined;
-  urlExposureCache = {};
-  const experimentStorageName = `EXP_${apiKey.slice(0, 10)}`;
-  let user;
-  try {
-    user = JSON.parse(
-      globalScope.localStorage.getItem(experimentStorageName) || '{}',
-    );
-  } catch (error) {
-    user = {};
-  }
+export class DefaultWebExperimentClient implements WebExperimentClient {
+  private readonly apiKey: string;
+  private readonly initialFlags: [];
+  private readonly config: WebExperimentConfig;
+  private readonly globalScope = getGlobalScope();
+  private readonly experimentClient: ExperimentClient | undefined;
+  private appliedInjections: Set<string> = new Set();
+  private appliedMutations: Record<string, MutationController[]> = {};
+  private previousUrl: string | undefined = undefined;
+  // Cache to track exposure for the current URL, should be cleared on URL change
+  private urlExposureCache: Record<string, Record<string, string | undefined>> =
+    {};
+  private flagVariantMap: Record<string, Record<string, Variant>> = {};
+  private localFlagKeys: string[] = [];
+  private remoteFlagKeys: string[] = [];
+  private isRemoteBlocking = false;
+  private customRedirectHandler: ((url: string) => void) | undefined;
+
+  constructor(
+    apiKey: string,
+    initialFlags: string,
+    config: WebExperimentConfig = {},
+  ) {
+    if (!this.globalScope || !isLocalStorageAvailable()) {
+      throw new Error(
+        'Amplitude Web Experiment Client could not be initialized.',
+      );
+    }
+    this.apiKey = apiKey;
+    this.initialFlags = JSON.parse(initialFlags);
+    // merge config with defaults and experimentConfig (if provided)
+    this.config = {
+      ...Defaults,
+      ...config,
+      ...(this.globalScope?.experimentConfig ?? {}),
+    };
 
-  // if web_exp_id does not exist:
-  // 1. if device_id exists, migrate device_id to web_exp_id and remove device_id
-  // 2. if device_id does not exist, create a new web_exp_id
-  // 3. if both device_id and web_exp_id exist, remove device_id
-  if (!user.web_exp_id) {
-    user.web_exp_id = user.device_id || UUID();
-    delete user.device_id;
-    globalScope.localStorage.setItem(
-      experimentStorageName,
-      JSON.stringify(user),
-    );
-  } else if (user.web_exp_id && user.device_id) {
-    delete user.device_id;
-    globalScope.localStorage.setItem(
-      experimentStorageName,
-      JSON.stringify(user),
-    );
-  }
+    if (this.globalScope?.webExperiment) {
+      return this.globalScope?.webExperiment;
```

Comment:
Given people should not be calling this function, we should pull this check outside the constructor